// @shaderfrog/glsl-parser/dist/preprocessor/preprocessor.test.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_1 = __importDefault(require("fs"));
var path_1 = __importDefault(require("path"));
var peggy_1 = __importDefault(require("peggy"));
var util_1 = __importDefault(require("util"));
var preprocessor_1 = require("./preprocessor");
var generator_1 = __importDefault(require("./generator"));
var fileContents = function (filePath) {
    return fs_1.default.readFileSync(path_1.default.join(__dirname, filePath)).toString();
};
var grammar = fileContents('preprocessor-grammar.pegjs');
var parser = peggy_1.default.generate(grammar, { cache: true });
var parse = function (src) { return parser.parse(src); };
var debugProgram = function (program) {
    debugAst(parse(program));
};
var debugAst = function (ast) {
    console.log(util_1.default.inspect(ast, false, null, true));
};
var expectParsedProgram = function (sourceGlsl) {
    var ast = parse(sourceGlsl);
    var glsl = (0, generator_1.default)(ast);
    if (glsl !== sourceGlsl) {
        debugAst(ast);
        expect(glsl).toBe(sourceGlsl);
    }
};
// test('pre test file', () => {
//   expectParsedProgram(fileContents('./preprocess-test-grammar.glsl'));
// });
test('preprocessor ast', function () {
    expectParsedProgram("\n#line 0\n#version 100 \"hi\"\n#define GL_es_profile 1\n#extension all : disable\n#error whoopsie\n#define A 1\nbefore if\n      #if A == 1 || B == 2\n      inside if\n      #define A\n          #elif A == 1 || defined(B) && C == 2\n          float a;\n          #elif A == 1 || defined(B) && C == 2\n          float a;\n      #define B\n      #endif\noutside endif\n#pragma mypragma: something(else)\nfinal line after program\n");
});
test('nested expand macro', function () {
    var program = "#define X Y\n#define Y Z\nX";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("Z");
});
test('if expression', function () {
    var program = "\n#define A\nbefore if\n#if !defined(A) && (defined(B) && C == 2)\ninside first if\n#endif\n#if ((defined(B) && C == 2) || defined(A))\ninside second if\n#endif\nafter if\n";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nbefore if\ninside second if\nafter if\n");
});
test('evaluate if branch', function () {
    var program = "\n#define A\nbefore if\n#if defined(A)\ninside if\n#endif\nafter if\n";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nbefore if\ninside if\nafter if\n");
});
test('evaluate elseif branch', function () {
    var program = "\n#define A\nbefore if\n#if defined(B)\ninside if\n#elif defined(A)\ninside elif\n#else\nelse body\n#endif\nafter if";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nbefore if\ninside elif\nafter if");
});
test('empty branch', function () {
    var program = "before if\n#ifdef GL_ES\nprecision mediump float;\n#endif\nafter if";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("before if\nafter if");
});
test('evaluate else branch', function () {
    var program = "\n#define A\nbefore if\n#if defined(D)\ninside if\n#elif defined(E)\ninside elif\n#else\nelse body\n#endif\nafter if";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nbefore if\nelse body\nafter if");
});
test('empty branch', function () {
    var program = "before if\n#ifdef GL_ES\nprecision mediump float;\n#endif\nafter if";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("before if\nafter if");
});
test('self referential object macro', function () {
    var program = "\n#define first first second\n#define second first\nsecond";
    // If this has an infinte loop, the test will never finish
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nfirst second");
});
test('self referential function macro', function () {
    var program = "\n#define foo() foo()\nfoo()";
    // If this has an infinte loop, the test will never finish
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nfoo()");
});
test('self referential macro combinations', function () {
    var program = "\n#define b c\n#define first(a,b) a + b\n#define second first(1,b)\nsecond";
    // If this has an infinte loop, the test will never finish
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\n1 + c");
});
test("function call macro isn't expanded", function () {
    var program = "\n#define foo() no expand\nfoo";
    var ast = parse(program);
    // debugAst(ast);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nfoo");
});
test("macro that isn't macro function call call is expanded", function () {
    var program = "\n#define foo () yes expand\nfoo";
    var ast = parse(program);
    // debugAst(ast);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\n() yes expand");
});
test('unterminated macro function call', function () {
    var program = "\n#define foo() yes expand\nfoo(\nfoo()";
    var ast = parse(program);
    expect(function () { return (0, preprocessor_1.preprocessAst)(ast); }).toThrow('foo( unterminated macro invocation');
});
test('macro function calls with no arguments', function () {
    var program = "\n#define foo() yes expand\nfoo()\nfoo\n()";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nyes expand\nyes expand");
});
test('macro function calls with bad arguments', function () {
    expect(function () {
        (0, preprocessor_1.preprocessAst)(parse("\n      #define foo( a, b ) a + b\n      foo(1,2,3)"));
    }).toThrow("'foo': Too many arguments for macro");
    expect(function () {
        (0, preprocessor_1.preprocessAst)(parse("\n      #define foo( a ) a + b\n      foo(,)"));
    }).toThrow("'foo': Too many arguments for macro");
    expect(function () {
        (0, preprocessor_1.preprocessAst)(parse("\n      #define foo( a, b ) a + b\n      foo(1)"));
    }).toThrow("'foo': Not enough arguments for macro");
});
test('macro function calls with arguments', function () {
    var program = "\n#define foo( a, b ) a + b\nfoo(x + y, (z-t + vec3(0.0, 1.0)))\nfoo\n(q,\nr)\nfoo(,)";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\nx + y + (z-t + vec3(0.0, 1.0))\nq + r\n + ");
});
test('nested function macro expansion', function () {
    var program = "\n#define X Z\n#define foo(x, y) x + y\nfoo (foo (a, X), c)";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\na + Z + c");
});
test('token pasting', function () {
    var program = "\n#define COMMAND(NAME)  { NAME, NAME ## _command ## x ## y }\nCOMMAND(x)";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast);
    expect((0, generator_1.default)(ast)).toBe("\n{ x, x_commandxy }");
});
test('preservation', function () {
    var program = "\n#line 0\n#version 100 \"hi\"\n#define GL_es_profile 1\n#extension all : disable\n#error whoopsie\n#define  A 1\nbefore if\n#if A == 1 || B == 2\ninside if\n#define A\n#elif A == 1 || defined(B) && C == 2\nfloat a;\n#define B\n#endif\noutside endif\n#pragma mypragma: something(else)\nfunction_call line after program";
    var ast = parse(program);
    (0, preprocessor_1.preprocessAst)(ast, {
        // ignoreMacro: (identifier, body) => {
        //   // return identifier === 'A';
        // },
        preserve: {
            conditional: function (path) { return false; },
            line: function (path) { return true; },
            error: function (path) { return true; },
            extension: function (path) { return true; },
            pragma: function (path) { return true; },
            version: function (path) { return true; },
        },
    });
    expect((0, generator_1.default)(ast)).toBe("\n#line 0\n#version 100 \"hi\"\n#extension all : disable\n#error whoopsie\nbefore if\ninside if\noutside endif\n#pragma mypragma: something(else)\nfunction_call line after program");
});
/*
test('debug', () => {
  const program = `
precision highp float;
precision mediump int;
precision lowp int;
`;

  const ast = parse(program);
  preprocessAst(ast);
  expect(generate(ast)).toBe(`
varying vec2 vUv;
`);
});
*/
