// vuetify/lib/composables/form.mjs
// Utilities
import { computed, inject, provide, ref } from 'vue';
import { useProxiedModel } from "./proxiedModel.mjs";
import { consoleWarn, propsFactory } from "../util/index.mjs"; // Types

export const FormKey = Symbol.for('vuetify:form');
export const makeFormProps = propsFactory({
  disabled: Boolean,
  fastFail: Boolean,
  lazyValidation: Boolean,
  readonly: Boolean,
  modelValue: {
    type: Boolean,
    default: null
  }
});
export function createForm(props) {
  const model = useProxiedModel(props, 'modelValue');
  const isDisabled = computed(() => props.disabled);
  const isReadonly = computed(() => props.readonly);
  const isValidating = ref(false);
  const items = ref([]);
  const errorMessages = ref([]);

  async function validate() {
    const results = [];
    let valid = true;
    errorMessages.value = [];
    model.value = null;
    isValidating.value = true;

    for (const item of items.value) {
      const itemErrorMessages = await item.validate();

      if (itemErrorMessages.length > 0) {
        valid = false;
        results.push({
          id: item.id,
          errorMessages: itemErrorMessages
        });
      }

      if (!valid && props.fastFail) break;
    }

    errorMessages.value = results;
    model.value = valid;
    isValidating.value = false;
    return {
      valid,
      errorMessages: errorMessages.value
    };
  }

  function reset() {
    items.value.forEach(item => item.reset());
    model.value = null;
  }

  function resetValidation() {
    items.value.forEach(item => item.resetValidation());
    errorMessages.value = [];
    model.value = null;
  }

  provide(FormKey, {
    register: (id, validate, reset, resetValidation) => {
      if (items.value.some(item => item.id === id)) {
        consoleWarn(`Duplicate input name "${id}"`);
      }

      items.value.push({
        id,
        validate,
        reset,
        resetValidation
      });
    },
    unregister: id => {
      items.value = items.value.filter(item => {
        return item.id !== id;
      });
    },
    isDisabled,
    isReadonly,
    isValidating,
    items
  });
  return {
    errorMessages,
    isDisabled,
    isReadonly,
    isValidating,
    items,
    validate,
    reset,
    resetValidation
  };
}
export function useForm() {
  return inject(FormKey, null);
}
